Adding expand_dims for xtensor

AllenDowney opened the pull request:
# Add expand_dims operation for labeled tensors

This PR adds support for the `expand_dims` operation in PyTensor's labeled tensor system, allowing users to add new dimensions to labeled tensors with explicit dimension names.

## Key Features

- New `ExpandDims` operation that adds a new dimension to an XTensorVariable
- Support for both static and symbolic dimension sizes
- Automatic broadcasting when size > 1
- Integration with existing tensor operations
- Full compatibility with xarray's expand_dims behavior

## Implementation Details

The implementation includes:

1. New `ExpandDims` class in `pytensor/xtensor/shape.py` that handles:
   - Adding new dimensions with specified names
   - Support for both static and symbolic sizes
   - Shape inference and validation

2. Rewriting rule in `pytensor/xtensor/rewriting/shape.py` that:
   - Converts labeled tensor operations to standard tensor operations
   - Handles broadcasting when needed
   - Validates symbolic sizes

3. Comprehensive test suite in `tests/xtensor/test_shape.py` covering:
   - Basic dimension expansion
   - Static and symbolic sizes
   - Error cases and edge cases
   - Compatibility with xarray operations
   - Integration with other labeled tensor operations

## Usage Example

```python
import pytensor.tensor as pt
from pytensor.xtensor import xtensor

# Create a labeled tensor
x = xtensor("x", dims=("city",), shape=(3,))

# Add a new dimension
y = expand_dims(x, "country")  # Adds a new dimension of size 1
z = expand_dims(x, "country", size=4)  # Adds a new dimension of size 4
```

## Testing

The implementation includes extensive tests that verify:
- Correct behavior with various input shapes
- Proper handling of symbolic sizes
- Error cases (invalid dimensions, sizes, etc.)
- Compatibility with xarray's expand_dims
- Integration with other labeled tensor operations


<!-- readthedocs-preview pytensor start -->
----
📚 Documentation preview 📚: https://pytensor--1449.org.readthedocs.build/en/1449/

<!-- readthedocs-preview pytensor end -->

AllenDowney commented:
Now that we have this PR based on the right commit, @ricardoV94 it is ready for a first look.

One question: my first draft of this was based on a later commit -- this draft goes back to an earlier commit, and it looks like `@register_xcanonicalize` doesn't exist yet, so I've replaced it with `@register_lower_xtensor`, which seems to be its predecessor. Is that the right thing to do for now?

ricardoV94 commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +382,75 @@ def squeeze(x, dim=None):
+
+    def __init__(self, dim, size=1):
+        self.dims = dim
+        self.size = size
```
This does not allow symbolic sizes, check UnStack for reference

AllenDowney commented:
@ricardoV94 I think this is a step toward handling symbolic sizes, but there are a couple of place where I'm not sure what the right behavior is. See the comments in `test_shape.py`, `test_expand_dims_implicit`.

Do those tests make sense? Are there more cases that should be covered?

ricardoV94 commented:
The simplest test for symbolic expand_dims is: 
```python
size_new_dim = xtensor("size_new_dim", shape=(), dtype=int)
x = xtensor("x", shape=(3,))
y =  x.expand_dims(new_dim=size_new_dim)
xr_function = function([x, size_new_dim], y)

x_test = xr_arange_like(x)
size_new_dim_test = DataArray(np.array(5, dtype=int))
result = xr_function(x_test, size_new_dim_test)
expected_result = x_test.expand_dims(new_dim=size_new_dim_test)
xr_assert_allclose(result, expected_result)
```
Yout can parametrize the test to try default and explicit non-default axis as well.

Sidenote, what is an implicit `expand_dims`? I don't think that's a thing.

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +380,195 @@ def test_squeeze_errors():
         fn2(x2_test)
+
+
+def test_expand_dims_explicit():
```
There's a lot of redundancy in these tests, nothing is gained from testing 1D, 2D, 3D...

ricardoV94 replied:
Let me push back on this argument. This test will run thousands to times in a year, small costs add up. For some people a nice test suite takes milliseconds to run for other under  a few hours. Speed is subjective

I can't conceive of a scenario where expand_dims would fail for 2d but work for 3d or vice versa. If so why not test up to nd?

My comment was also a bit more broad, expand_dims is not such a complex operation that I think we need to test to exhausition. Right now I think it's the Op with most checks after indexing (which is actually really complex). 

We also know AI suggestions tend to err on the side of verbosity not conciseness and I suppose that translates to the tests

ricardoV94 commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +382,75 @@ def squeeze(x, dim=None):
+    if not isinstance(dim, str):
+        raise TypeError(f"`dim` must be a string or None, got: {type(dim)}")
+
+    if dim in x.type.dims:
+        raise ValueError(f"Dimension {dim} already exists in {x.type.dims}")
```
this check should be done in `make_node`

ricardoV94 commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +382,75 @@ def squeeze(x, dim=None):
+    if dim in x.type.dims:
+        raise ValueError(f"Dimension {dim} already exists in {x.type.dims}")
+
+    if isinstance(size, int | np.integer):
+        if size <= 0:
+            raise ValueError(f"size must be positive, got: {size}")
+    elif not (
+        hasattr(size, "ndim") and getattr(size, "ndim", None) == 0  # symbolic scalar
+    ):
+        raise TypeError(f"size must be an int or scalar variable, got: {type(size)}")
+
+    # Always convert size to a PyTensor scalar variable
+    size_var = as_tensor(size, ndim=0)
```
This kind of conversion/check should be done in `make_node

ricardoV94 commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +382,75 @@ def squeeze(x, dim=None):
+    __props__ = ("dims",)
+
+    def __init__(self, dim):
+        self.dims = dim
```
Nit: If the Op only allows one dim use singular, if it allows multiple use plural name. 

Important: Do not create `__init__` with a different signature than the variable name you end up saving into.

ricardoV94 replied:
Also feel free to do any validation here, must dim be a string? Check for that

AllenDowney replied:
I see now that `xarray.DataArray.expand_dims` allows the `dim` parameter to be "hashable, sequence of hashable, dict, or None "

Currently we are handling only string and None.

Should we generalize to handle all cases handled by xarray?

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -301,6 +303,15 @@ def test_squeeze_explicit_dims():
     fn3d = xr_function([x3], y3d)
     xr_assert_allclose(fn3d(x3_test), x3_test)
 
+    # Reversibility with expand_dims
```
Still don't think this test is needed

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +380,195 @@ def test_squeeze_errors():
+
+    # Duplicate dimension creation
+    y = expand_dims(x, "new")
+    with pytest.raises(ValueError):
```
match the expected error messages to be sure you are triggering the branch you care about. Sometimes you are testing an earlier error and can't tell because of only checking for ValueError/TypeError

AllenDowney commented:
@ricardoV94 I've addressed most of your comments on the previous round, and made a first pass at adding support for multiple dimensions. Please take a look at the `expand_dims` wrapper function, which canonicalizes the inputs and loops through them to make a series of Ops.

Assuming that adding multiple dimensions is rare, what do with think of the loop option, as opposed to making a single Op that adds multiple dimensions?

ricardoV94 commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +383,115 @@ def squeeze(x, dim=None):
+    def make_node(self, x, size):
+        x = as_xtensor(x)
+
+        if not isinstance(self.dim, str):
```
This can be done already in init

ricardoV94 commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +383,115 @@ def squeeze(x, dim=None):
+
+        if self.dim in x.type.dims:
+            raise ValueError(f"Dimension {self.dim} already exists in {x.type.dims}")
+        if isinstance(size, int | np.integer):
+            if size <= 0:
+                raise ValueError(f"size must be positive, got: {size}")
+        elif not (
+            hasattr(size, "ndim")
+            and getattr(size, "ndim", None) == 0  # symbolic scalar
+        ):
+            raise TypeError(
+                f"size must be an int or scalar variable, got: {type(size)}"
+            )
+
+        # Convert size to tensor
+        size = as_tensor(size, ndim=0)
```
Better to convert to as_xtensor, since both tensor and xtensor scalars will be accepted then.

Instead of having the check for numpy specifically, check if after conversion you have `isinstance(size, Constant) and (size.data <= 0)` and then raise.

Are you sure xarray doesn't support size of zero? It's valid otherwise in numpy

ricardoV94 commented on pytensor/xtensor/rewriting/shape.py:
```diff
@@ -132,3 +139,32 @@ def local_squeeze_reshape(fgraph, node):
+    x, size = node.inputs
+    out = node.outputs[0]
+    # Lower to tensor.expand_dims(x, axis=0)
+    from pytensor.tensor import expand_dims as tensor_expand_dims
```
import globally

ricardoV94 commented on pytensor/xtensor/rewriting/shape.py:
```diff
@@ -132,3 +139,32 @@ def local_squeeze_reshape(fgraph, node):
+    expanded = Assert(msg="size must be positive")(expanded, gt(size, 0))
+    # If size is not 1, broadcast
+    try:
+        static_size = get_scalar_constant_value(size)
```
You already inferred static size in make_node so you can use that info from `out.type.shape`

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,224 @@ def test_squeeze_errors():
+
+    # Test behavior with symbolic size > 1
+    # NOTE: This test documents our current behavior where expand_dims broadcasts to the requested size.
+    # This differs from xarray's behavior where expand_dims always adds a size-1 dimension.
```
I don't understand the point. The test shows xarray accepts `expand_dims({dim_name: size})` and I guess also `expand_dims(dim_name=4)`. That's exactly the behavior we want to replicate. If the size kwarg is something that doesn't exist in xarray we shouldn't introduce it, we want to mimick their API

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    )
+    xr_assert_allclose(fn(xr_arange_like(x)), expected)
+
+    # Insert new dim between existing dims
```
I suppose this test is meant to cover the axis argument? As it stands it doesn't make sense, it's testing transpose explicitly which is tested elsewhere

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    expected = x_test.expand_dims("new").transpose("a", "new", "b")
+    xr_assert_allclose(fn(x_test), expected)
+
+    # Expand with multiple dims
```
I suppose this test is intended to test multiple dims in a single call instead?

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    size_sym_1 = scalar("size_sym_1", dtype="int64")
+    x = xtensor("x", dims=("a", "b"), shape=(2, 3))
+    y = expand_dims(x, "batch", size=size_sym_1)
+    fn = xr_function([x, size_sym_1], y, on_unused_input="ignore")
```
The unused input is evidence the test is wrongly specified ;)

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    fn = xr_function([x], y)
+    x_test = xr_arange_like(xtensor(dims=x.dims, shape=(2, 3, 5)))
+    res = fn(x_test)
+    expected = x_test.expand_dims({"d": 3})  # 3 is the size of dimension "b"
```
Instead of the comment explanation write the expected to mimick the Pytensor expression 
```suggestion
    expected = x_test.expand_dims({"d": x_test.sizes["b"]})
```

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    expected = x_test.expand_dims({"d": 3})  # 3 is the size of dimension "b"
+    xr_assert_allclose(res, expected)
+
+    # Test broadcasting with symbolic size from a different tensor
```
Don't see the point of this case

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    # expected = x_test.expand_dims("batch")  # always size 1
+    # xr_assert_allclose(res, expected)
+
+    # Test using symbolic size from a reduction operation
```
Don't see the point of this test

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    expected = x_test.expand_dims({"batch": 3})  # 3 is the size of dimension "b"
+    xr_assert_allclose(res, expected)
+
+    # Test chaining expand_dims with symbolic sizes
```
These tests are trying really hard to not just use a symbolic size variable, why?

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    expected = x_test.expand_dims({"b": 2}).expand_dims({"c": 2})
+    xr_assert_allclose(res, expected)
+
+    # Test bidirectional broadcasting with symbolic sizes
```
Don't see the point of this test

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    size_sym_1 = scalar("size_sym_1", dtype="int64")
+    size_sym_2 = scalar("size_sym_2", dtype="int64")
+    y = expand_dims(x, {"country": size_sym_1, "state": size_sym_2})
+    fn = xr_function([x, size_sym_1, size_sym_2], y, on_unused_input="ignore")
```
On unused input is suspect

ricardoV94 commented on pytensor/xtensor/rewriting/shape.py:
```diff
@@ -132,3 +134,35 @@ def local_squeeze_reshape(fgraph, node):
+        result_tensor = expand_dims(x_tensor, new_axis)
+    else:
+        # First expand with size 1
+        expanded = expand_dims(x_tensor, new_axis)
```
Small eager optimization, since the new axis is always on the left you don't need the expand_dims step and can skip directly to boradcast_to.

ricardoV94 commented:
Rewrite looks good. As we discussed we should redo the tests to use expand_dims as a method (like xarray users would).

I suspected xarray allows specifying the size like `x.expand_dims(dim_a=1, dim_b=2)` which is equivalent to `x.expand_dims({"dim_a":1, "dim_b":2})`. At least that was a pattern I noticed in other xarray methods.

I saw you had a test for multiple dims with dict, but I didn't see one with kwargs.

ricardoV94 commented on tests/xtensor/test_shape.py:
```diff
@@ -369,3 +371,213 @@ def test_squeeze_errors():
+    y = expand_dims(x, "batch", size=size_sym_1)
+    fn = xr_function([x, size_sym_1], y, on_unused_input="ignore")
+    x_test = xr_arange_like(x)
+    xr_assert_allclose(fn(x_test, 1), x_test.expand_dims("batch"))
```
Since you already have the function you can use it twice (makes comment before stale)
```suggestion
    xr_assert_allclose(fn(x_test, 1), x_test.expand_dims("batch"))
    xr_assert_allclose(fn(x_test, 5), x_test.expand_dims({"batch": 5}))
```

AllenDowney commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +384,119 @@ def squeeze(x, dim=None):
+        if self.dim in x.type.dims:
+            raise ValueError(f"Dimension {self.dim} already exists in {x.type.dims}")
+
+        # Check if size is a valid type before converting
+        if not (
+            isinstance(size, int | np.integer)
+            or (hasattr(size, "ndim") and getattr(size, "ndim", None) == 0)
+        ):
+            raise TypeError(
+                f"size must be an int or scalar variable, got: {type(size)}"
+            )
+
+        # Determine shape
+        try:
+            static_size = get_scalar_constant_value(size)
+        except NotScalarConstantError:
+            static_size = None
+
+        if static_size is not None:
+            new_shape = (int(static_size), *x.type.shape)
+        else:
+            new_shape = (None, *x.type.shape)  # symbolic size
+
+        # Convert size to tensor
+        size = as_xtensor(size, dims=())
+
+        # Check if size is a constant and validate it
+        if isinstance(size, Constant) and size.data < 0:
+            raise ValueError(f"size must be 0 or positive, got: {size.data}")
```
It's cleaner, but it also break two tests -- some of the complexity of the current version came from fixing these errors. How much debugging time do we want to spend on cleaning this up?

AllenDowney replied:
I'm not sure how to answer that other than pasting in the results of the test run.

AllenDowney commented:
@ricardoV94 I cleaned up the code as suggested and took a first cut at handling the `axis` parameter. Please take a look.

ricardoV94 commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +384,112 @@ def squeeze(x, dim=None):
+    for name, size in dims_dict.items():
+        canonical_dims.append((name, size))
+
+    # Store original dimensions for later use with axis
+    original_dims = list(x.type.dims)
```
This isn't used other than to copy it below, just define it then directly?
```suggestion
```

AllenDowney replied:
We  need to store the original dimensions. The reason is:
We add new dimensions at the front
Then we need to know where the original dimensions were to correctly place the new ones at their specified axis positions
If we used x.type.dims after adding the new dimensions, we'd get duplicates because the new dimensions are already in the tensor

AllenDowney replied:
Right.

ricardoV94 commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +384,112 @@ def squeeze(x, dim=None):
+        # Start with original dims, then insert each new dim at its axis
+        target_dims = list(original_dims)
+        # axis values are relative to the result after each insertion
+        for insert_dim, insert_axis in sorted(
+            zip(new_dim_names, axis), key=lambda x: x[1]
+        ):
```
Maybe you can use np.insert? You may also want to normalize the axis before you get here if they are negative. There's a helper in npy2_compat.py

AllenDowney replied:
I cleaned this up and it now works with negative axis parameters

ricardoV94 reviewed (commented):
This looks great. Just the question about the size kwarg and small notes.

AllenDowney commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +384,112 @@ def squeeze(x, dim=None):
+        dims_dict = {}
+        for name, val in dim.items():
+            if isinstance(val, Sequence | np.ndarray) and not isinstance(val, str):
+                dims_dict[name] = len(val)
```
yes

AllenDowney commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +383,120 @@ def squeeze(x, dim=None):
+    # Extract size from dim_kwargs if present
+    size = dim_kwargs.pop("size", 1) if dim_kwargs else 1
+
+    # xarray compatibility: error if a sequence (list/tuple) of dims and size are given
+    if (isinstance(dim, list | tuple)) and ("size" in locals() and size != 1):
+        raise ValueError("cannot specify both keyword and positional arguments")
```
yup.

AllenDowney commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +383,120 @@ def squeeze(x, dim=None):
+    elif isinstance(dim, dict):
+        dims_dict = {}
+        for name, val in dim.items():
+            if isinstance(val, Sequence | np.ndarray) and not isinstance(val, str):
```
That's an error, but I've added a check to catch it.

ricardoV94 commented on pytensor/xtensor/shape.py:
```diff
@@ -380,3 +383,120 @@ def squeeze(x, dim=None):
+            "create_index_for_new_dim=False has no effect in pytensor.xtensor",
+            UserWarning,
+            stacklevel=2,
+        )
```
I think that's good
